### ISSUE TITLE
[Feature Request] Compatibility between Sky Rendering mods

### ISSUE THREAD

**Issue by Abastro** [MinecraftForge/MinecraftForge#3665]
Currently, there is only way to mod the sky: rendering the whole sky in alternative IRenderHandler.
But there are several cases that partial modification of the sky is needed, which suffers incompatibility  issue in current system.

Here are some examples of the cases:
- Recoloring the sun/moon/stars/the whole sky.
- Rendering additional celestial objects on the sky.
- Changing the way of the sky and the celestial system rendered. (Or, shaders)

Also, referencing certain celestial object is needed for some mods, to modify and augment the object.

I suggest this system to solve this issue:
First solution:

- Celestial Object Registry or similar, to manage the objects. It is collection of objects which is registered and managed here.
- Modification Layer, to allow recoloring (or color multiplication). Event system may be better here.
- Renderer, to allow specific rendering style of the sky. Shaders can be applied here.

Second solution:

 - Categorize whole rendering into Near / Sky / Planets (Sun, Moon & Others) / Stars.
 - Give special case for the Sun and the Moon, so that their color can be adjusted.

I want to get feedbacks on this.

**Comment by Daomephsta** [comment]
What does this allow that wrapping the existing sky renderer doesn't?

**Comment by Abastro** [comment]
Compatibility. It's possible to just replace whole sky rendering, but I saw many cases which only need partial changes. This can help most of the incompatibility issues with the sky (celestial objects).

**Comment by Daomephsta** [comment]
By wrapping I mean creating a renderer like this:
```java
    public class WrappedRenderer implements IRenderHandler
   {
	private final IRenderHandler prevHandler;
	
	public WrappedRenderer(IRenderHandler prevHandler)
	{
	    this.prevHandler = prevHandler;
	}
	
	public void render(float partialTicks, WorldClient world, Minecraft mc);
	{
	    //Call the wrapped renderer's render method
	    if(prevHandler != null) prevHandler.render(partialTicks, world, mc);
	    //Now render your stuff
	}
    }
```
This allows partial changes. If you wanted to say add an extra sun, you'd put the necessary code after the call to prevHandler.render(). The previous handler would render the sky and then your extra sun would be rendered on top.

**Comment by Abastro** [comment]
Still, it does not allow to e.g. change moon's color.

**Comment by r4j4h** [comment]
True, but we could add a data payload to the call chain to carry that information. Participating RenderHandlers could read from the data payload to advise their rendering and then write to it to inform others of what they rendered or what their state is (moon position, color, etc).

E.g, We could extend `IRenderHandler` with a different render function we could pass a generic `Map` or something through and have our RenderHandlers optionally store/fetch from that.

For use with vanilla, one could stub the vanilla renderer as a reference implementation that populates the map with all of the stats you mention, e.g. sun position, etc. and would use values in the map if they were already there. Pass it a map with 'moonColor' to set the moon color to what you want, or read the map's moonColor after prevHandler to see what it set it to.

---

Striving for compatibility between skyboxes is a great ideal, but we need to consider ways to remain as flexible as possible. And how it interacts with vanilla.

For example, changing the moon's color is great, but what if someone else adds another moon, how do we control each in unison and independently?

`moon2Color` or `moons[2].color` come to mind first but don't feel like they are scaling right. Maybe namespacing solves it: `minecraft:moon.color` for vanilla, and let others pick their own scheme under their namespace. Feels good and aligns with how compatibility is with other aspects, but can we introspect at runtime and learn about what's there?

You mentioned some data points like brightness and color. Lots of numbers inter-relate and may or may not be useful to different consumers. Position/Angle of objects changes depending on coordinate system, and while it's great to say stick with renderer since that's the current base, if things are doing logic then they may want the source algorithm numbers. E.g. screen space vs object space vs world space. It may be a potential resource strain if we don't provide the right numbers they need in the pay load and lots of people end up doing the same trig conversions all the time. Color can do this too if they are going RGB to HSL and back. I think we need to get these more concrete so we can evaluate between them more effectively.

What if they don't want celestial bodies for their skybox? Like what if they want it to look like the insides of a giant worm? Still may have information for making different areas animate or change color or texture, but now isn't even remotely trying to simulate a celestial body. But maybe they do want to control how sun tracking devices behave in there? Or can let it occur and ignore it - assuming the abstraction never leaks?

Alternatively, a superlarge segment of space with ships flying over but no real celestial movement around a distinct center (many bodies but each's rotations are from different origin points) and maybe the station is mainly what you see but through shader-generated reflections the other bodies do interesting light play through shaders on the distant parts of the station. They may want data about traffic levels of space ships flying overhead through some lanes or activity on the ship like lights or vents or smoke which is different but may also go on schedules. Even still they may want to reuse parts of distant galaxies/stars twinkling from the normal renderer. And manipulate where sensors think the sun/moon are, etc.

Even still, maybe it's a dimensional mod and the player has interested a demonic spooky dimension where the sun trackers are purposely wrong to illustrate how far out this new level is. They can see the sun that way but the tracker points the other way.

Compatibility is a hard goal if everyone isn't making their skyboxes with the flexibility of the system or use cases in mind. I think it may help to make this feature request a reality to hammer down these goals further with use cases and concrete examples of the color, positioning, shape, and other data that is desired in the system.

---

Because of the work involved I am thinking this may be two related but distinct features: **a sky rendering pipeline with introspection and modification built in** and a **compatible implementation that replicates vanilla minecraft's skybox system**.

Leviathan's example starts down the pipeline without introducing anything for your introspection/control needs, but could with some changes. My example isn't the only way. A mod that does only this intended to be required by mods that want to do fun skyboxy things would be just as good a fit as Forge I think.

Modifying vanilla's skybox code for the control/info you want is a different level of work that would be awesome if it were done and I imagine it already has many times by individuals. This time it needs to be made to work with that pipeline and should only have to be done once and then re-used (potentially multiple times) by the pipeline.

A mod that recreates what vanilla does but integrated with that pipeline would be good, but in practice it would be overwriting the true vanilla code.

Having Forge do this does feel like the 'better' spot as vanilla would be built-in and patched, but this necessitates Forge also provides the Pipeline as well as it makes no sense for forge to depend or come with the pipeline mod.

I think having mods take the responsibility involves less risk for Forge and while there'd be risk of deviation from vanilla, multiple attempts could be made and evaluated and battle-tested by players of the mods and would avoid an all-in-one patch-fest on like three big vanilla functions like if Forge has to do it.

tl;dr:

I'm no contributor here (yet?) but I like this idea a lot and want to see it happen, but am waffling on if Forge is the right home for it and I think we need to work on the specifications of this request more either way.

**Comment by mezz** [comment]
This feature sounds like it is of specific interest to a smallish number of mods, and might be better served by integration between those mods directly.

For instance, Buildcraft, Forestry, EnderIO, (and probably others) have a concept of coolants and integrate together to agree on coolant values of various liquids. There doesn't really need to be a formal coolant compatibility layer in Forge to support it.

**Comment by Abastro** [comment]
@mezz the difference is that it's nearly impossible to implement the compatibility with sky manually. Coolants are basically blocks/items/fluids which is well-managed, and compatibility can simply be solved by registry. On the other hand, celestial object(or skybox object) is not even defined by any means. Thus hard to get any compromises. Providing compatibility layer from an individual mod will be hard with significant performance drop.

Now I'm quite sure that this one is related with the limit of minecraft and ignorance on the sky. Okay then, I'll accept that this suggestion is meaningless.

**Comment by mezz** [comment]
Oh I didn't realize you wanted to alter the vanilla sky directly, I thought you just wanted compatibility between modded skies.

**Comment by Abastro** [comment]
There's already two celestial objects on vanilla, though: The Sun and the Moon. It's hardly a modification on vanilla.
Many mods add certain effects based on them, modifying their colors, but the rendering can only be accessed with sky renderer. Again, it's pain to implement the compatibility manually here.

**Comment by williewillus** [comment]
still valid, currently you can only replace the whole sky rendering (which leads to a ton of copied code). Being able to replace things one by one would be better.

**Comment by Abastro** [comment]
Now I think sky rendering should be at least divided into categories, which could be registrable.
Like stars / planets(sun&moon&others) / sky.
One can either overwrite/add additional rendering on each category.
Then some render events for sun/moon/others about their colors can solve the other problem.  (ColorCelestialEvent with specific celestial object ids? Or just for sun/moon would be enough)

Any other thoughts on sky rendering?
